gin/config_parser.py

```diff
@@ -437,6 +452,42 @@ def _maybe_parse_configurable_reference(self):
 
     return True, reference
 
+  def _maybe_parse_identifier_reference(self):
+    """Try to parse a identifier reference([package.package.]var_or_fn_or_cls_name[()])"""
+    location = self._current_location()
+
+    if self._current_token.kind != tokenize.NAME:
+      self._raise_syntax_error('Unexpected token.')
+
+    begin_line_num = self._current_token.begin[0]
+    begin_char_num = self._current_token.begin[1]
+    end_char_num = self._current_token.end[1]
+    line = self._current_token.line
+
+    identifier_parts = []
+    # This accepts an alternating sequence of NAME  or '.' tokens.
+    step_parity = 0
+    while (step_parity == 0 and self._current_token.kind == tokenize.NAME or
+           step_parity == 1 and self._current_token.value =='.'):
+      identifier_parts.append(self._current_token.value)
+      step_parity = not step_parity
+      end_char_num = self._current_token.end[1]
+      self._advance_one_token()
+    self._skip_whitespace_and_comments()
+    identifier = ''.join(identifier_parts)
+    untokenized_identifier = line[begin_char_num:end_char_num]
+    valid_format = MODULE_RE.match(identifier)
+    if untokenized_identifier != identifier or not valid_format:
+      location = (self._filename, begin_line_num, begin_char_num + 1, line)
+      self._raise_syntax_error('Malformatted identifier.', location)
+
+    self._skip_whitespace_and_comments()
+
+    with utils.try_with_location(location):
+      reference = self._delegate.identifier_reference(identifier)
```

Comment:
Should we report error here if the identifier cannot be found? Otherwise it will be delayed until when the identifier reference is actually deepcopied when the a configurable function is called, which can be hard to understand.

Reply:
we have no information about an identifier at the stage of `parsing` , it's been evaluated at the context frame (eg: local variable in a function)

Reply:
I think those symbols should have been imported by the time it's reference in the gin file, right? Similar to ConfigurableReference.

Reply:
No , it's different with `ConfigurableReference` ,  we can refer to any identifier, just need to make sure it's available in the scope where function that ref to it is called 

eg : 
```python
  def testLocalVariableReference(self):
    config_str = """
          configurable2.non_kwarg = a
    """
    config.parse_config(config_str)

    a = 1   
    instance, _ = configurable2()  #  
    self.assertEqual(instance, a)
```
`a`  is local variable ,  defined before `configurable2` is called

Reply:
> Can you also submit a PR to ALF about installing this new version of gin after being merged?

ok

Reply:
> No , it's different with `ConfigurableReference` , we can refer to any identifier, just need to make sure it's available in the scope where function that ref to it is called
> 
> eg :
> 
> ```python
>   def testLocalVariableReference(self):
>     config_str = """
>           configurable2.non_kwarg = a
>     """
>     config.parse_config(config_str)
> 
>     a = 1   
>     instance, _ = configurable2()  #  
>     self.assertEqual(instance, a)
> ```
> 
> `a` is local variable , defined before `configurable2` is called

What is the use case for this scenario? It doesn't seem to belong to configuration.

Reply:
it's just a test case for usage,  
`a` is an identifier , it can be a `local variable` , `function` or `class` 
eg : `torch.nn.relu` , `numpy.pi`, `dict` ...